Update SPIRE Upstream Authority plugin to poll bundle updates

marcosy opened the pull request:
Signed-off-by: Marcos G. Yedro <marcosyedro@gmail.com>

<!--
1. If this is your first PR, please read our contributor guidelines
https://github.com/spiffe/spiffe/blob/master/CONTRIBUTING.md
https://github.com/spiffe/spire/blob/master/CONTRIBUTING.md
-->

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included? (not yet)
- [x] Documentation updated?

**Affected functionality**
SPIRE Upstream Authority plugin

**Description of change**
Make SPIRE Upstream Authority plugin to poll for bundle changes and update streams when changes are detected.

**Which issue this PR fixes**
Fixes: #1452


I still need to update/add unit tests but I'm pushing this PR as draft to get some early feedback and validate the approach.

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -70,59 +88,141 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
-	ctx := stream.Context()
+func (m *Plugin) pollBundleUpdates(ctx context.Context) {
+	ticker := time.NewTicker(upstreamPollFreq)
+	for {
```
same suggestion here about moving the code inside of the `<-ticker.C` case above the select so that the initial bundle fetching happens right away.

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -70,59 +88,141 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
+		select {
+		case <-ticker.C:
+			newRootCAs := m.getBundle().RootCas
+			if !areRootsEqual(rootCAs, newRootCAs) {
```
We have to be careful here. The first time through the roots could compare equal since the bundle may not have been fetched yet (both will be empty). If we're relying on the roots not being equal to send the first response, this may not do that.

More critically, we have to make sure the roots returned here is as fresh or fresher than the roots returned by the `submitCSRUpstreamCA` call.

marcosy replied:
Good point! I'll update `m.currentBundle` with the rootsCA contained in the `submitCSRUpstreamCA` response.

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire_workload.go:
```diff
@@ -5,32 +5,59 @@ import (
+
 		case <-ctx.Done():
-			return nil, nil, nil, ctx.Err()
+			m.log.Debug("watch workload context done", "reason", ctx.Err())
```
Should the `nodeClient` and `conn` be cleaned up before returning here?

marcosy replied:
Right!

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -70,59 +88,141 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
+		select {
+		case <-ticker.C:
+			newKeys := m.getBundle().JwtSigningKeys
+			if !arePublicKeysEqual(keys, newKeys) {
```
We have to be careful here. The first time through the keys could compare equal since the bundle may not have been fetched yet (both will be empty). If we're relying on the keys not being equal to send the first response, this may not do that.

More critically, we have to make sure the keys here is as fresh or fresher than the keys returned by the `PushJWTKeyUpstream` call.

marcosy replied:
Good point! I'll update `m.currentBundle` with the keys contained in the `PushJWTKeyUpstream` response.

marcosy commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -70,59 +88,141 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
-	m.mtx.RLock()
-	defer m.mtx.RUnlock()
-	ctx := stream.Context()
+	m.subscribeToPolling()
```
My intention with `subscribeToPolling` is to keep only one set of:
- workload API client
- node client
- poller thread
And share them for all stream RPCs, as long as at least one RPC is active.

I believe the workload API would block too in the previous code if it does not get an update

marcosy commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -70,59 +88,141 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
+	defer m.unsubscribeToPolling()
 
-	wCert, wKey, wBundle, err := m.getWorkloadSVID(ctx, m.config)
+	m.nodeMtx.RLock()
+	certChain, _, err := submitCSRUpstreamCA(stream.Context(), m.nodeClient, request.Csr)
+	m.nodeMtx.RUnlock()
```
I totally agree! Fixing...

marcosy commented:
Thanks for the feedback! I addressed all the comments and updated the current tests to have some coverage. I still need to add more tests to cover all scenarios.

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -89,121 +89,118 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
-	m.nodeMtx.RLock()
-	certChain, _, err := submitCSRUpstreamCA(stream.Context(), m.nodeClient, request.Csr)
-	m.nodeMtx.RUnlock()
+	certChain, err := m.submitCSRUpstreamCA(stream.Context(), request.Csr)
```
This still seems racy. Consider the sequence of events:
1. Fetch loop grabs bundle with roots `A`
1. `submitCSRUpstreamCA` takes the bundle lock, gets a new CA and gets roots 'B', and updates the cached roots to `B`
1. Fetch loop finally gets the bundle lock and updates bundle with `A`
1. loop below picks up changes `A` and sends them down as a bundle. `B` is never sent.

The key takeaway is that only bundle content fetched _after_ the upstream RPC has finished minting the downstream CA or publishing the JWT key should be considered.

There are a few ways to implement this.

One approach is that there is a counter that is bumped whenever the X.509 roots are updated. The fetch loop could record the counter before it fetches. After it gets a new bundle, it could make sure the roots have not been updated out from underneath it by checking the counter. If the roots have been updated, then the roots in the bundle are ignored. A similar counter could be used for JWT keys.

marcosy replied:
Oh, this scenario is being considered. Fetch loop fetches and updates the bundle within the bundle lock at: `fetchAndSetBundle`.

The same occurs with MintX509 in `submitCSRUpstreamCA` and with PublishJWTKey in: `pushAndSetInitialKeys`

azdagron replied:
Ah, i missed that. TBH, I'm not in love with holding locks over RPC calls in general.

marcosy replied:
Oh I see the point of that, I'll change it for the counter approach so we can remove that lock.

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire_node.go:
```diff
@@ -15,20 +15,33 @@ import (
-func submitCSRUpstreamCA(ctx context.Context, nodeClient node.NodeClient, csr []byte) ([]*x509.Certificate, *bundleutil.Bundle, error) {
-	resp, err := nodeClient.FetchX509CASVID(ctx, &node.FetchX509CASVIDRequest{
+func (m *Plugin) submitCSRUpstreamCA(ctx context.Context, csr []byte) ([]*x509.Certificate, error) {
+	m.bundleMtx.Lock()
```
Do we want separate mutex for updating X.509 roots v.s. JWT keys?

marcosy replied:
I thought about that... My only concern was that having too many mutexes in the struct could have a negative impact in readability/maintainability... but perhaps it's not that bad

azdagron replied:
I think it's fine with a single mutex considering the current layout.

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -29,10 +39,22 @@ func BuiltIn() catalog.Plugin {
+	nodeMtx    sync.RWMutex
+	nodeClient node.NodeClient
+	conn       *grpc.ClientConn
+	log        hclog.Logger
```
nit: `log` shouldn't be grouped with the fields protected by `nodeMtx`

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -29,10 +39,22 @@ func BuiltIn() catalog.Plugin {
+	conn       *grpc.ClientConn
+	log        hclog.Logger
+
+	pollMtx                sync.RWMutex
```
nit: not using the read portion of the lock, change to `sync.Mutex` (more self-documenting)

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -29,10 +39,22 @@ func BuiltIn() catalog.Plugin {
+	currentPollSubscribers uint64
+
+	bundleMtx     sync.RWMutex
+	bundleVersion uint
```
nit: How would you feel about making this explicitly 64-bit? Even though it will be 64-bit on 64-bit architectures, I'd rather be worried about overflow on a 32-bit platform.
```suggestion
	bundleVersion uint64
```

azdagron replied:
Of course, overflow wouldn't really matter now that I think of it. But it still sets off my spidey sense (32-bit counters, oh my!)

marcosy replied:
sure, will change it 🕷️  👍

azdagron commented on pkg/server/plugin/upstreamauthority/spire/spire.go:
```diff
@@ -70,59 +92,166 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
-	if err != nil {
-		return err
+	keys := []*common.PublicKey{}
+	ticker := time.NewTicker(internalPollFreq)
```
nit: can the speed of the unit tests be improved by having these tickers use the clock as well?

azdagron replied:
I'm ok with this being addressed in a follow up PR. We'd like this in for 0.10.0 and that release timeline trumps a few additional seconds in a unit test.

marcosy replied:
Perfect, I'll let that improvement for another PR
